modules/projections/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/projections/Message.scala

```diff
@@ -0,0 +1,96 @@
+package ch.epfl.bluebrain.nexus.sourcing.projections
+
+import akka.persistence.query.{EventEnvelope, NoOffset, Offset}
+import cats.Functor
+import ch.epfl.bluebrain.nexus.sourcing.projections.ProjectionProgress.ProgressStatus
+import ch.epfl.bluebrain.nexus.sourcing.projections.ProjectionProgress.ProgressStatus.Failed
+
+/**
+  * A message with context information. The context information provides the progress status.
+  *
+  * @param progress       map where each key is a progressId and each value is its status
+  * @param value          the value wrapped in the message
+  * @param currProgressId the current progress Id
+  * @param offset         the offset of the message
+  * @param persistenceId  the persistenceId of the message
+  * @param sequenceNr     the sequence number of the message
+  * @tparam A the generic type of the message value
+  */
+final case class Message[A](
+    progress: Map[String, ProgressStatus],
+    value: A,
+    currProgressId: String,
+    offset: Offset,
+    persistenceId: String,
+    sequenceNr: Long
+) {
+
+  /**
+    * Changes the ''currProgressId'' to the passed value
+    *
+    * @param progressId the new ''currProgressId''
+    * @return a copy of the message with the ''currProgressId'' as the provided ''progressId''
+    */
+  def nextProgress(progressId: String): Message[A] =
+    copy(
+      progress = progress + (progressId -> progress.getOrElse(progressId, ProgressStatus.Passed)),
+      currProgressId = progressId
+    )
+
+  def unit: Message[Unit] =
+    copy(value = ())
+
+  /**
+    * Adds failed to the current message progress.
+    *
+    * @return a copy of the current message with failed added to the progress map
+    */
+  def fail(error: String): Message[A] =
+    copy(progress = progress + (currProgressId -> ProgressStatus.Failed(error)))
+
+  /**
+    * Adds discard to the current message progress. Discard won't be added if the current progress is Failed.
+    *
+    * @return a copy of the current message with discard added to the progress map
+    */
+  def discard(): Message[A] =
+    if (progress.getOrElse(currProgressId, ProgressStatus.Passed).failed) this
+    else copy(progress = progress + (currProgressId -> ProgressStatus.Discarded))
+
+  /**
+    * Combine two messages progress
+    */
+  def +[B](that: Message[B]): Message[B] =
```

Comment:
rename to `addProgressOf` since the most important field of a message is its value